### PR TITLE
Tag Prefix 2nd Try

### PR DESCRIPTION
**Pull request by patricegautier**
Hi Flavio,

Thanks for the detailed feedback.   All of it made great sense..

I think I caught everything in this new pull request, and added a couple of test cases as well.

Let me know!

### CODE REVIEW

File: internal/cmd/check.go
Comment (flavio) [inline internal/cmd/check.go]: I would do something like that:
```go
msg := fmt.Sprintf(
   "%s is already the latest version available that satisfies the %s constraint", 
   evaluation.Image,
   evaluation.Constraint)
if evaluation.TagPrefix != "" {
   msg = fmt.Sprintf("%s and the tag prefix %s", msg, evaluation.TagPrefix)
}
fmt.Println(msg)
```

File: internal/cmd/check.go
Comment (flavio) [inline internal/cmd/check.go]: remove the `string` type from here, you can inherit the one from `configFile`

File: internal/cmd/check.go
Comment (flavio) [inline internal/cmd/check.go]: Same here, but this time remove the `string` type from the left

File: internal/workers/job_handler.go
Comment (flavio) [inline internal/workers/job_handler.go]: remove the `string` type from the left

File: pkg/fresh_container/client.go
Comment (flavio) [inline pkg/fresh_container/client.go]: remove the `string` type from the left

File: pkg/fresh_container/comparer.go
Comment (flavio) [inline pkg/fresh_container/comparer.go]: remove the string type on the left

File: pkg/fresh_container/comparer.go
Comment (flavio) [inline pkg/fresh_container/comparer.go]: please run gofmt against all the files you changed, there are other similar issues around

File: pkg/fresh_container/image.go
Comment (flavio) [inline pkg/fresh_container/image.go]: remove the string type from the left

File: pkg/fresh_container/image.go
Comment (flavio) [inline pkg/fresh_container/image.go]: remove the empty line

File: pkg/fresh_container/tags.go
Comment (flavio) [inline pkg/fresh_container/tags.go]: remove the empty line

File: pkg/fresh_container/tags.go
Comment (flavio) [inline pkg/fresh_container/tags.go:25]: I think this new implementation is overly complicated, I would just do:

```golang
func TagsToVersions(tags []string, tagPrefix string, skipInvalid bool) (versions semver.Versions, err error) {
   for _, tag := range tags {
                tag = strings.TrimPrefix(tag, tagPrefix)
		v, err := semver.Parse(tag)
		if err == nil {
			versions = append(versions, v)
		} else if skipInvalid {
			log.WithFields(log.Fields{
				"tag":   tag,
                                "tagPrefix": tagPrefix,
				"error": err}).Warn("Skipping image tag")
		} else {
			return semver.Versions{}, err
		}
	}

	return versions, nil
}
```

**Review by flavio — CHANGES_REQUESTED** [review]
Hi, thanks for the fixes you've made.

I've left some comments, please do not open a new PR, but rather send the fixes as new commits into this PR.

You can then squash everything into a single commit as a last step, once I've approved all the changes.

Please, configure your IDE/editor to run gofmt on save, there are many small things that could be fixed by it.

File: pkg/fresh_container/tags.go
Comment (patricegautier) [reply pkg/fresh_container/tags.go:25]: indeed.

**Comment by patricegautier** [comment]
Thanks again for the great feedback. 

Take a look I think it's close.

**Comment by patricegautier** [comment]
Well hang on - the server case needs work too..

**Comment by patricegautier** [comment]
I had to bang my head on bit on gorilla/mux but it works now..

**Review by flavio — CHANGES_REQUESTED** [review]
Yes, this looks definitely better :clap: 

You still have to fix the background job definition, as you pointed out. I'm looking forward to see the last fix :)

File: cmd/fresh-container/main.go
Comment (flavio) [inline cmd/fresh-container/main.go:97]: ~~Not really, the code has been changed so that it processes all the tags, regardless of the prefix they use. If they have a prefix that matches the one provided here, this will be stripped~~

Sorry, I didn't read the code later on. What you stated is correct

File: internal/api/v1/check.go
Comment (flavio) [inline internal/api/v1/check.go]: You should really automate `gofmt` ;)

File: internal/api/v1/server.go
Comment (flavio) [inline internal/api/v1/server.go]: nitpicking, remove the empty line

File: internal/api/v1/server.go
Comment (flavio) [inline internal/api/v1/server.go:55]: Yeah, that looks like the only solution... I did some research too :(

**Review by flavio — CHANGES_REQUESTED** [review]
Last round of changes. Please run `gofmt`, otherwise the next time someone with `gofmt` enabled changes one of these files he will get a list of unrelated changes

File: internal/api/v1/check.go
Comment (patricegautier) [reply internal/api/v1/check.go]: grrr I thought I had

File: internal/api/v1/server.go
Comment (patricegautier) [reply internal/api/v1/server.go:55]: I could see this being a real pain if you need a few optional query values.. 2^n handlers!

File: internal/api/v1/check.go
Comment (patricegautier) [reply internal/api/v1/check.go]: Ok I see the problem - I missed a set of directories when I set it up..

**Comment by patricegautier** [comment]
The background job to build the container on the GitHub side is failing.  I am going to need some help to get unblocked here I think:

https://github.com/patricegautier/fresh-container/runs/3968310010?check_suite_focus=true

it's failing with:

#30 exporting to image
483
#30 pushing layers 0.9s done
484
#30 ERROR: unexpected status: 403 Forbidden
485
------
486
 > exporting to image:
487
------
488
error: failed to solve: unexpected status: 403 Forbidden
489
Error: buildx failed with: error: failed to solve: unexpected status: 403 Forbidden

EDIT:  Looks like it might be a resource problem:  https://github.com/docker/build-push-action/issues/238

**Comment by patricegautier** [comment]
There's a permission problem of some sort preventing the build of the container from succeeding, having to do with the image workflow.  I tried tweaking container-image.yml but that's not enough..

**Comment by patricegautier** [comment]
also, it seems that having a hard-coded reference to the Flavio/fresh-container tag in gets in the way of forking.. There's probably a variable to solve that too.

File: .github/workflows/container-image.yml
Comment (flavio) [inline .github/workflows/container-image.yml]: Let's leave this change out of this PR, it's unrelated

File: .github/workflows/container-image.yml
Comment (flavio) [inline .github/workflows/container-image.yml]: same here

**Review by flavio — CHANGES_REQUESTED** [review]
LGTM.

Before I can merging it, please do the following actions:

  * Remove the changes to the GitHub Action file, we will tackle that with another PR
  * Squash all the commits together, then force push. If you don't know how to do that I can help

Thanks again for your contribution

**Comment by flavio** [comment]
JFYI: I fixed the GitHub issue with https://github.com/flavio/fresh-container/commit/59a14bba5db76cba73158559db00f6a136331f1e, it's already inside of the `main` branch

**Comment by patricegautier** [comment]
Ok I reverted the container name change in container-image.yml

I do need instructions on how to squash and force-push though..

**Comment by flavio** [comment]
Oh, no worries about the squash and merge. I see GH added an option for doing that on my side (as someone who has merge rights). I'll handle that

**Comment by flavio** [comment]
I've merged everything as you can see, I'll release a new version in the next days

**Comment by patricegautier** [comment]
Great, thank you!

**Pull request merged**
